tools/dds/dds-sniffer/rs-dds-sniffer.cpp

```diff
@@ -181,88 +208,175 @@ void dds_sniffer::on_writer_removed( librealsense::dds::dds_guid guid, const cha
     remove_topic_writer( guid, topic_name );
 }
 
-void dds_sniffer::on_reader_added( librealsense::dds::dds_guid guid, const char* topic_name )
+void dds_sniffer::on_reader_added( librealsense::dds::dds_guid guid, const char * topic_name )
 {
-    if (_print_discoveries)
+    if( _print_discoveries )
     {
         print_reader_discovered( guid, topic_name, true );
     }
     save_topic_reader( guid, topic_name );
 }
 
-void dds_sniffer::on_reader_removed( librealsense::dds::dds_guid guid, const char* topic_name )
+void dds_sniffer::on_reader_removed( librealsense::dds::dds_guid guid, const char * topic_name )
 {
-    if (_print_discoveries)
+    if( _print_discoveries )
     {
         print_reader_discovered( guid, topic_name, false );
     }
 
     remove_topic_reader( guid, topic_name );
 }
 
-void dds_sniffer::on_participant_added( librealsense::dds::dds_guid guid, const char* participant_name )
+void dds_sniffer::on_participant_added( librealsense::dds::dds_guid guid, const char * participant_name )
 {
-    if (_print_discoveries)
+    if( _print_discoveries )
     {
         print_participant_discovered( guid, participant_name, true );
     }
 
-    std::lock_guard<std::mutex> lock( _dds_entities_lock );
+    std::lock_guard< std::mutex > lock( _dds_entities_lock );
 
     _discovered_participants[guid] = participant_name;
 }
-void dds_sniffer::on_participant_removed( librealsense::dds::dds_guid guid, const char* participant_name )
+
+void dds_sniffer::on_participant_removed( librealsense::dds::dds_guid guid, const char * participant_name )
 {
     if( _print_discoveries )
     {
         print_participant_discovered( guid, participant_name, false );
     }
 
-    std::lock_guard<std::mutex> lock( _dds_entities_lock );
+    std::lock_guard< std::mutex > lock( _dds_entities_lock );
 
     _discovered_participants.erase( guid );
 }
 
-void dds_sniffer::save_topic_writer( librealsense::dds::dds_guid guid, const char* topic_name )
+void dds_sniffer::on_type_discovery( char const * topic_name, DynamicType_ptr dyn_type )
+{
+    // Register type with participant
+    TypeSupport type_support( new DynamicPubSubType( dyn_type ) );
+    type_support.register_type( _participant.get() );
+    std::cout << "Discovered topic " << topic_name << " of type: " << type_support->getName() << std::endl;
```

Comment:
How does this differ from the message printed in `print_writer_discovered`?

Reply:
The message here is about the discovered topic, that can be registered in the domain without any readers/writers using it. A topic in a domain is identified by it's name and data type, here we output both.
The message in `print_writer_discovered` is about the writer entity, supplying some information about it like it's GUID and the topic it publishes.

Reply:
OK, so:
* Does this function get called even for topics without XTypes?
* Can an explanation be added in comments at the start of this function?
* Can you please send us a sample output of the debug log?

Reply:
* No, this function is called only for topics with TypeObject
* I will add before the function a comment about it being called only for X-Types that their TypeObject was received
* A sample of the output with debug log sent to console
Note that there is logging from dds-participant when debug log is active and a similar print to screen from rs-dds-sniffer that is always active

 25/07 14:54:13,836 DEBUG [24012] (dds-participant.cpp:134) participant 'rs-dds-sniffer' is up on domain 0
rs-dds-sniffer listening on domain 0
 25/07 14:54:18,002 DEBUG [23816] (dds-participant.cpp:36) participant 'rs-dds-server' discovered
Participant 01.0f.d0.39.80.41.c1.d9.01.00.00.00|0.0.1.c1 rs-dds-server_4180 discovered
 25/07 14:54:18,002 WARNING [14768] (rs-dds-sniffer.cpp:51) [DDS] Buffer is being invalidated, segment_size may be insufficient
 25/07 14:54:18,577 DEBUG [17148] (dds-participant.cpp:57) new DataWriter (01.0f.d0.39.80.41.c1.d9.01.00.00.00|0.0.1.3) publishing 'realsense/device-info' of type 'librealsense::dds::topics::raw::device_info'
DataWriter  01.0f.d0.39.80.41.c1.d9.01.00.00.00|0.0.1.3 publishing topic 'realsense/device-info' discovered
 25/07 14:54:18,587 DEBUG [17148] (dds-participant.cpp:97) Discovered topic realsense/device-info of type: librealsense::dds::topics::raw::device_info
Discovered topic realsense/device-info of type: librealsense::dds::topics::raw::device_info
 25/07 14:54:18,587 DEBUG [17148] (dds-participant.cpp:77) new DataReader (01.0f.d0.39.60.34.30.b8.01.00.00.00|0.0.1.4) reading topic 'realsense/device-info' of type 'librealsense::dds::topics::raw::device_info'
DataReader  01.0f.d0.39.60.34.30.b8.01.00.00.00|0.0.1.4 reading topic 'realsense/device-info' discovered
 25/07 14:54:18,601 DEBUG [17148] (rs-dds-sniffer.cpp:379) Subscriber matched
 25/07 14:54:18,601 WARNING [14768] (rs-dds-sniffer.cpp:51) [DDS] Attempting to add existing writer
Received topic realsense/device-info of type librealsense::dds::topics::raw::device_info
name: [I, n, t, e, l,  , R, e, a, l, S, e, n, s, e,  , D, 4, 0, 5,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ]
serial_number: [1, 0, 6, 1, 2, 2, 2, 7, 0, 3, 3, 2,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ]
product_line: [D, 4, 0, 0,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ]
topic_root: [r, e, a, l, s, e, n, s, e, /, D, 4, 0, 5, /, 1, 0, 6, 1, 2, 2, 2, 7, 0, 3, 3, 2,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,  ]
locked: false

Reply:
Have you noticed the two warnings from DDS?

Reply:
Yes, and there are 2 other messages that I didn't copy here
 31/07 09:32:36,717 WARNING [12180] (rs-dds-sniffer.cpp:52) [DDS] (ID:6992) Port 7412 Zombie. Reset the port
 31/07 09:32:36,733 WARNING [12180] (rs-dds-sniffer.cpp:52) [DDS] (ID:6992) Port 7413 Zombie. Reset the port

I believe that "Port XXX Zombie. Reset the port" and "Buffer is being invalidated, segment_size may be insufficient" are because we exit the applications with Ctrl+C with some unclosed DDS resources.

- Ports 7412 and 7413 are default port for meta-traffic unicast for domain 0, first participant. "Port XXX Zombie. Reset the port" appear for each domain that the sniffer previously ran in, even after restart, so I think there are some files that indicate it is open.
- "Buffer is being invalidated, segment_size may be insufficient" is issued when trying to allocate shared memory buffer for the first time (with discovery message) and seeing that buffers where allocated on a previous run and are unused anymore.

I am not fully sure why "Attempting to add existing writer" is issued, I think it is because we use both shared memory and TCPv4 transports on same host (It is issued for a topic discovery message)